storage/shard.go

```diff
@@ -180,6 +180,20 @@ type dbShardEntry struct {
 	index          uint64
 	curReadWriters int32
 	reverseIndex   struct {
+		// NB(prateek): writeAttemptedBlockstart[1-2]Nanos is used to indicate the index block start(s)
```

Comment:
I'm really leaning towards using a RW lock here. It'll be 24 more bytes per shardEntry, but will so greatly reduce the complexity of the code that it's a no brainer. Only finished writing this version up to convince myself it's a terrible idea.

Reply:
@robskillington please could you give this a once over and see if i'm missing something more obvious.